feat(service): add cap service

titouv opened the pull request:
## Changes
- Add template for the [Cap](https://cap.so/) service, a open-source Loom alternative

titouv committed:
cap.so deployment template first base

titouv commented:
I am waiting for the release of an ARM version of their image to fully test this template. Until then, the PR remains a draft

ShadowArcanist commented:
Hey @titouv 

Your SERVICE_FQDN variable won't work here because the container listens on port 3000 not on 5679.

I have a working compose for this:
```yaml
services:
  cap-web:
    container_name: cap-web
    image: 'ghcr.io/capsoftware/cap-web:latest'
    restart: unless-stopped
    environment:
      - SERVICE_FQDN_CAP_3000
      - 'DATABASE_URL=mysql://$SERVICE_USER_MYSQL:$SERVICE_PASSWORD_MYSQL@cap-db:3306/${MYSQL_DATABASE:-planetscale}'
      - 'WEB_URL=${WEB_URL:-http://localhost:3000}'
      - 'NEXTAUTH_URL=${WEB_URL:-http://localhost:3000}'
      - 'DATABASE_ENCRYPTION_KEY=${SERVICE_PASSWORD_64_DATABASEENCRYPTIONKEY}'
      - 'NEXTAUTH_SECRET=${SERVICE_PASSWORD_64_NEXTAUTHSECRET}'
      - 'CAP_AWS_ACCESS_KEY=${SERVICE_USER_MINIO}'
      - 'CAP_AWS_SECRET_KEY=${SERVICE_PASSWORD_MINIO}'
      - 'CAP_AWS_BUCKET=${CAP_AWS_BUCKET:-capso}'
      - 'CAP_AWS_REGION=${CAP_AWS_REGION:-us-east-1}'
      - 'CAP_AWS_ENDPOINT=${CAP_AWS_ENDPOINT:-http://minio:3902}'
    depends_on:
      cap-db:
        condition: service_healthy
      cap-s3:
        condition: service_healthy
        
  cap-db:
    container_name: cap-db
    image: 'mysql:8.0'
    restart: unless-stopped
    environment:
      - 'MYSQL_USER=${SERVICE_USER_MYSQL}'
      - 'MYSQL_PASSWORD=${SERVICE_PASSWORD_MYSQL}'
      - 'MYSQL_DATABASE=${MYSQL_DATABASE:-planetscale}'
      - 'MYSQL_ROOT_PASSWORD=${SERVICE_PASSWORD_MYSQLROOT}'
      - 'MYSQL_ALLOW_EMPTY_PASSWORD=${MYSQL_ALLOW_EMPTY_PASSWORD:-yes}'
    command:
      - '--max_connections=1000'
      - '--default-authentication-plugin=mysql_native_password'
    healthcheck:
      test:
        - CMD
        - mysqladmin
        - ping
        - '-h'
        - 127.0.0.1
      interval: 10s
      timeout: 10s
      retries: 5
    volumes:
      - 'cap_db:/var/lib/mysql'
  cap-s3:
    container_name: cap-s3
    image: 'bitnami/minio:latest'
    restart: unless-stopped
    environment:
      - MINIO_API_PORT_NUMBER=3902
      - MINIO_CONSOLE_PORT_NUMBER=3903
      - MINIO_ROOT_USER=$SERVICE_USER_MINIO
      - MINIO_ROOT_PASSWORD=$SERVICE_PASSWORD_MINIO
    healthcheck:
      test:
        - CMD
        - curl
        - '-f'
        - 'http://localhost:3902/minio/health/live'
      interval: 10s
      timeout: 10s
      retries: 5
    volumes:
      - 'cap_s3:/bitnami/minio/data'
```

Note: I built my own ARM image to test this and was able to reach the login page. However, I didn't proceed with further testing because the CAP desktop app keeps crashing whenever I try to record the screen, so I had to stop there.

titouv committed:
upgrade from comment

titouv commented:
Thanks for your help !

I also built a ARM image from the version 0.3.51 of cap (which is the same as the current public build of the desktop app). I was unable to make upload work when running the docker compose from the cap repo locally.

I updated the PR with your comment. I am waiting for the next releases to test upload again locally

peaklabs-dev committed:
Merge branch 'next' into v4.x

pievalentin commented:
Would love to see cap added. How could I help on this? I dont have a arm box but i could spin one in the cloud

peaklabs-dev commented:
@titouv For now please finish up the template with `platform: linux/amd64` as waiting for `arm64` images is not realistic as that can take months in some cases.

titouv committed:
updates with modification from cap docker-compose.yml example

titouv committed:
fixes for automatic bucket creation

titouv commented:
Hi,

The Cap.so team published an [arm image](https://github.com/CapSoftware/Cap/pkgs/container/cap-web).

I updated the template to use some new env var from cap.so docker-compose [template](https://github.com/CapSoftware/Cap/blob/main/docker-compose.template.yml)

I was just stuck when deploying locally, the upload functionality was not working, with the fixes in the lasts commits it should be good.

This was tested on a ARM VPS running Coolify [v4.0.0-beta.420.6](https://github.com/coollabsio/coolify/releases/tag/v4.0.0-beta.420.6)

I modified the PR state

pievalentin commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,80 @@
+      - 'S3_PUBLIC_ENDPOINT=${SERVICE_FQDN_CAPS3_3902}'
+      - 'S3_INTERNAL_ENDPOINT=${SERVICE_FQDN_CAPS3_3902}'
+      - 'NEXT_RUNTIME=nodejs'
+      - 'S3_PATH_STYLE=true'
```
If I want to use a s3 or a r2 bucket, I should point those to the bucket? I would be surprised if the storage on a regular VPS would support more than 20 videos

titouv replied:
This is was the cap team recommends on their [self-hosting docs](https://cap.so/docs/self-hosting). They do some heavy compressing by default but I agree i may be a bit shorts for VPS with not enough storage.

How do you want to proceed ? Have two docker-compose templates (like for some service with either integrated database or not), one with MinIO for local storage and the second with remove s3-compatible adresses ? Or remove this one with local storage if its not pertinent

pievalentin replied:
I don't know how they usually proceed. Personally, I would use R2 directly. @peaklabs-dev might know better. 
Some template have different versions for different DBs. I think only one template but with a doc or comment to explain how to connect to a remote bucket would make the most sense idk

titouv replied:
I added an small comments to explain that its possible to connect to a remote R2/S3 directly by removing cap-s3 and changing some env vars in 8ae2e2ef716209e72503dd43ea2bd41b5ee9e67c

pievalentin replied:
looks good would be happy to try

peaklabs-dev replied:
Normally, just ENVs are good. If you want minio, you can deploy another service (mino) in the same network and connect to it. Could you please change that (remove minio) and just add ENVs to fill out for the user?

titouv replied:
Just pushed a commit (3a81331) that changes the following things as you advised:

- Removed MinIO service completely
- Added comments on how to use different options by putting correct envs (deploying MinIO on the same Coolify instance, or using remote object storage with S3/R2)

pievalentin replied:
I think having a default minio would have made more sense. I can see how some less tech savvy users would be surprise the template doesn't not work out of the box. Your call @peaklabs-dev

Cinzya replied:
@titouv it is better if you open a PR in the coolify docs and add those instructions there. Not every user will view the compose file, hence why it is prefered use env variables over hardcoded values so it's not necessary to edit the compose file.

If Cap doesn't work out of the box without a s3 bucket, it might be good to mark those as required, then the user sees a warning message and can't deploy without filling those env variables. You can do that by adding a `:?` after the variable in the compose. E.g.:
```
- 'CAP_AWS_BUCKET=${CAP_AWS_BUCKET:?}'
```

titouv replied:
Thanks for the help, didn't know i could do that perfect I will open a PR

And thanks for the tips of `:?`, is there a doc page that indicate that this is possible ?

Cinzya replied:
we just added it into the docs, still needs to be pushed to the "live/production" version. It has been possible for a while now, but it was never documented. you can read more about it here: https://next.coolify.io/docs/knowledge-base/docker/compose#required-environment-variables

titouv committed:
add explanation for storage s3 local/remote storage option

titouv committed:
remove minio, put instructions to use env in  comments

pievalentin commented:
I just tried the raw docker-compose and the login need to use resend to send magic link. (https://cap.so/docs/self-hosting) or the user has to go in the logs to find the magic link. This should be mentioned as well.

Plus I tried to connect with the desktop app @titouv but I get a `ERR_INVALID_URL`  from the server. I suspect the FQDN envvar arent well set. Did you manage to connect the desktop app? if yes, what are your env vars like

titouv commented:
@pievalentin 

For the first point, I completely agree. I will add some indication for that in the documentation page as Cinzya pointed out.

For the second point, I am not sure if I can reproduce it, but for now, I have another issue that may be linked. When deploying a service with Docker Compose, the link generated for the domain connection is not valid. You can see in the screenshot that the domains generated do not include `http://` but just start at the domain. So the link generated in the interface is not well binded. It happens to me on this project but also when deploying `beszel.yaml` template, so this may be related to a change in Coolify (my version is Coolify v4.0.0-beta.426)

If the issue you have is not this one, can you add some details about how to reproduce it? I will investigate.

<img width="558" height="226" alt="SCR-20250901-kzlq" src="https://github.com/user-attachments/assets/beca7903-778e-41f3-8405-b46edad4cc1e" />

<img width="1018" height="613" alt="SCR-20250901-kzjc" src="https://github.com/user-attachments/assets/c08e2c08-2282-42a7-94e5-594a75c68a31" />

Cinzya commented:
> When deploying a service with Docker Compose, the link generated for the domain connection is not valid. You can see in the screenshot that the domains generated do not include `http://` but just start at the domain. So the link generated in the interface is not well binded. It happens to me on this project but also when deploying `beszel.yaml` template, so this may be related to a change in Coolify (my version is Coolify v4.0.0-beta.426)

That is due to recent changes to the magic env variables from the recent update.

You need to replace all `SERVICE_FQDN` variables with the `SERVICE_URL` and vise versa. They now behave the opposite way.

Cinzya commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+    image: 'ghcr.io/capsoftware/cap-web:latest'
+    restart: unless-stopped
+    environment:
+      - SERVICE_FQDN_CAP_3000
```
Every new service now needs to use the `SERVICE_URL` instead of the `SERVICE_FQDN` since the latest release.
```suggestion
      - SERVICE_URL_CAP_3000
```

Cinzya commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+    environment:
+      - SERVICE_FQDN_CAP_3000
+      - 'DATABASE_URL=mysql://$SERVICE_USER_MYSQL:$SERVICE_PASSWORD_MYSQL@cap-db:3306/${MYSQL_DATABASE:-planetscale}'
+      - 'WEB_URL=${SERVICE_FQDN_CAP_3000}'
+      - 'NEXTAUTH_URL=${SERVICE_FQDN_CAP_3000}'
```
```suggestion
      - 'WEB_URL=${SERVICE_URL_CAP_3000}'
      - 'NEXTAUTH_URL=${SERVICE_URL_CAP_3000}'
```

titouv committed:
SERVICE_FQDN to SERVICE_URL from @Cinzya

Co-authored-by: Cynthia Ebert <54354036+Cinzya@users.noreply.github.com>

titouv committed:
SERVICE_FQDN to SERVICE_URL from @Cinzya

Co-authored-by: Cynthia Ebert <54354036+Cinzya@users.noreply.github.com>

titouv committed:
mandatory fields for s3 storage

pievalentin commented:
I fixed my issue by hard-coding the FQDN for `WEB_URL` and `NEXTAUTH_URL` probably the fix with the service env var would have resolved this.

A non related item:
1. Do you know @titouv how can someone increase the number of seats? Is the cap team intentionally leaving this feature behind for the self-hosted version?
<img width="2330" height="900" alt="image" src="https://github.com/user-attachments/assets/140b0934-f4b7-4d46-9523-0da6ec5b6340" />

I looked in their codebase and could not find a dev flag somewhere. `NODE_ENV=dev` did not help

pievalentin commented:
> > When deploying a service with Docker Compose, the link generated for the domain connection is not valid. You can see in the screenshot that the domains generated do not include `http://` but just start at the domain. So the link generated in the interface is not well binded. It happens to me on this project but also when deploying `beszel.yaml` template, so this may be related to a change in Coolify (my version is Coolify v4.0.0-beta.426)
> 
> That is due to recent changes to the magic env variables from the recent update.
> 
> You need to replace all `SERVICE_FQDN` variables with the `SERVICE_URL` and vise versa. They now behave the opposite way.

<img width="2002" height="427" alt="image" src="https://github.com/user-attachments/assets/8a2157fd-de33-461d-8005-97cbfa3ce904" />

I think we should use the one without the `_3000` ?
Out of the box, the auth redirect is incorrect

ShadowArcanist commented:
https://github.com/coollabsio/coolify-docs/pull/352 - Just sending this here so I can see the status of this PR on Docs repo 🙂

titouv commented:
> 1. Do you know @titouv how can someone increase the number of seats? Is the cap team intentionally leaving this feature behind for the self-hosted version?
> 
> <img alt="image" width="2000" height="900" src="https://private-user-images.githubusercontent.com/14977219/484322225-140b0934-f4b7-4d46-9523-0da6ec5b6340.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY5MzMwOTEsIm5iZiI6MTc1NjkzMjc5MSwicGF0aCI6Ii8xNDk3NzIxOS80ODQzMjIyMjUtMTQwYjA5MzQtZjRiNy00ZDQ2LTk1MjMtMGRhNmVjNWI2MzQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAzVDIwNTMxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhN2IzMjMyY2M1OTAwZDM1YWYxMDZlZjQwODkwYjhlMmVhY2NmYWZhZDM5YTNkMjU0MzZmOGUzZTlmOTk1YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7FPcHjc5j69VzKR1tGAg24giptLJef7pvJmeIuvPn7U">
> I looked in their codebase and could not find a dev flag somewhere. `NODE_ENV=dev` did not help

From my understanding the easiest way i found when self hosting is updating the DB manually. Maybe there is a user facing way that the cap team enabled for self-hosting but i'm not sure.

Here is a screenshot with an example of how you can do that
<img width="2133" height="412" alt="SCR-20250903-ujht" src="https://github.com/user-attachments/assets/49011036-2967-4e59-ae3e-c33b7cbb3e38" />

pievalentin commented:
> > 1. Do you know @titouv how can someone increase the number of seats? Is the cap team intentionally leaving this feature behind for the self-hosted version?
> > 
> > <img alt="image" width="2000" height="900" src="https://private-user-images.githubusercontent.com/14977219/484322225-140b0934-f4b7-4d46-9523-0da6ec5b6340.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY5MzMwOTEsIm5iZiI6MTc1NjkzMjc5MSwicGF0aCI6Ii8xNDk3NzIxOS80ODQzMjIyMjUtMTQwYjA5MzQtZjRiNy00ZDQ2LTk1MjMtMGRhNmVjNWI2MzQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAzVDIwNTMxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhN2IzMjMyY2M1OTAwZDM1YWYxMDZlZjQwODkwYjhlMmVhY2NmYWZhZDM5YTNkMjU0MzZmOGUzZTlmOTk1YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7FPcHjc5j69VzKR1tGAg24giptLJef7pvJmeIuvPn7U">
> > I looked in their codebase and could not find a dev flag somewhere. `NODE_ENV=dev` did not help
> 
> From my understanding the easiest way i found when self hosting is updating the DB manually. Maybe there is a user facing way that the cap team enabled for self-hosting but i'm not sure.
> 
> Here is a screenshot with an example of how you can do that <img alt="SCR-20250903-ujht" width="2000" height="412" src="https://private-user-images.githubusercontent.com/39600279/485351092-49011036-2967-4e59-ae3e-c33b7cbb3e38.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwMTgyOTYsIm5iZiI6MTc1NzAxNzk5NiwicGF0aCI6Ii8zOTYwMDI3OS80ODUzNTEwOTItNDkwMTEwMzYtMjk2Ny00ZTU5LWFlM2UtYzMzYjdjYmIzZTM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA0VDIwMzMxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViNjE5ZDE4ZTNhMmExZGEzMWY3YWQ1OWNiY2E0MjdhYzNmMTQwYzc5MzM4MzNhYjc1OTFlMGE2MmU4YzQ2MDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KHAB94G3QZz3Njzp9eXj9TI6BJZbRSeN4qiOVAvF6Tc">

I will try that! I know they need to make a living, but I am surprised no one fixed that upstream

ShadowArcanist commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+      - SERVICE_URL_CAP_3000
+      - 'DATABASE_URL=mysql://$SERVICE_USER_MYSQL:$SERVICE_PASSWORD_MYSQL@cap-db:3306/${MYSQL_DATABASE:-planetscale}'
+      - 'WEB_URL=${SERVICE_URL_CAP_3000}'
+      - 'NEXTAUTH_URL=${SERVICE_URL_CAP_3000}'
```
Replace `SERVICE_URL_CAP_3000` with `SERVICE_URL_CAP` for WEB and NEXTAUTH URLs because we currently have a bug affecting some services env which has port number at the end - https://github.com/coollabsio/coolify/issues/6670#issuecomment-3322126426

The below suggestion fixes the app generating wrong login redirect url

```suggestion
      - 'WEB_URL=${SERVICE_URL_CAP}'
      - 'NEXTAUTH_URL=${SERVICE_URL_CAP}'
```

ShadowArcanist commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+  cap-web:
+    container_name: cap-web
+    image: 'ghcr.io/capsoftware/cap-web:latest'
+    restart: unless-stopped
```
Coolify adds this automatically

```suggestion
```

ShadowArcanist commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+
+services:
+  cap-web:
+    container_name: cap-web
```
Coolify automatically gives container names as <service_name>-<service_uuid>
<img width="852" height="229" alt="image" src="https://github.com/user-attachments/assets/f61a3ea2-4ba1-4c95-a018-22607e27881e" />


```suggestion
```

ShadowArcanist commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+        condition: service_healthy
+        
+  cap-db:
+    container_name: cap-db
```
Coolify automatically gives container names as <service_name>-<service_uuid> as mentioned earlier
```suggestion
```

ShadowArcanist commented on templates/compose/cap.yaml:
```diff
@@ -0,0 +1,79 @@
+  cap-db:
+    container_name: cap-db
+    image: 'mysql:8.0'
+    restart: unless-stopped
```
Coolify adds this automatically
```suggestion
```

ShadowArcanist reviewed (changes requested):
Hey @titouv Please implement the requested changes and include your steps to update the database to increase seat counts.

On Cap's discord server I found a suggestion to set stripe status on database so dashboard and cap desktop app will show user have pro subscription, it would be great if you could include steps to do this on the docs PR

<img width="838" height="311" alt="image" src="https://github.com/user-attachments/assets/cde911a9-9d26-4db1-bf5e-a43284e9532e" />

titouv committed:
Update templates/compose/cap.yaml

Co-authored-by: ShadowArcanist <162910371+ShadowArcanist@users.noreply.github.com>

titouv committed:
Update templates/compose/cap.yaml

Co-authored-by: ShadowArcanist <162910371+ShadowArcanist@users.noreply.github.com>

titouv committed:
Update templates/compose/cap.yaml

Co-authored-by: ShadowArcanist <162910371+ShadowArcanist@users.noreply.github.com>

titouv committed:
Update templates/compose/cap.yaml

Co-authored-by: ShadowArcanist <162910371+ShadowArcanist@users.noreply.github.com>

titouv committed:
Update templates/compose/cap.yaml

Co-authored-by: ShadowArcanist <162910371+ShadowArcanist@users.noreply.github.com>

titouv commented:
Hi @ShadowArcanist, thanks a lot for your review !
I accepted your changes and added the specific instructions to unlock the limits (org seats and recordings) to the coolify-docs PR https://github.com/coollabsio/coolify-docs/pull/352

andrasbacsai commented:
Thank you for the PR!
